### PR TITLE
New resource `r/azurerm_virtual_desktop_host_pool_registration_info` to enable new computed value of hostpool registration token

### PR DESCRIPTION
**Pull request by amccullough84**
There is an issue with the current implementation of the virtual_desktop_host_pool resource which limits usability.

When a host pool is created, a registration token is required to register any session hosts to the hostpool. This is performed by providing a valid registration token and hostpool name to a VM extension which enables the association.

The registration token is part of the registration_info block in the resource and the value is computed. The token is generated based on the deployment providing an expiration date. When this is provided or changed a new registration token is generated.

This works fine for the first apply, and subsequent runs until the token expires at which point it is no longer possible to add session hosts to the host pool.

As the value of the registration token is computed, it does not refresh the value in state when the expiration date in the configuration is changed until a refresh occurs prior to the next apply. In order to resolve is it currently necessary to maintain separate configuration for configuration of the host pool and deployment of session hosts and to perform a terraform refresh before retrieving the output from the hostpool registration info which can then be passed as a variable to the session host deployment configuration.

Using a CustomizeDiff and SetNewComputed would be a great solution, but unfortunately this is not possible on a nested attribute (unless I've missed a creative way to achieve this?).

The workaround I've implemented is to use CustomizeDiff and SetNewComputed to refresh the computed value of the token as a new top-level attribute - registration_token.

While this differs from the API implementation, the lack of ability to refresh the registration token without splitting the configuration and running terraform refresh in between or running an apply twice and expecting the first time to fail is a major limitation.

Acceptance Test Passed:

=== RUN   TestAccVirtualDesktopHostPool_update
=== PAUSE TestAccVirtualDesktopHostPool_update
=== CONT  TestAccVirtualDesktopHostPool_update
--- PASS: TestAccVirtualDesktopHostPool_update (187.07s)
PASS

Test configuration:

```
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
    }
  }
}

provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-vdesktop-1"
  location = "uksouth"
}

resource "azurerm_virtual_desktop_host_pool" "test" {
  name                 = "acctestHP"
  location             = azurerm_resource_group.test.location
  resource_group_name  = azurerm_resource_group.test.name
  type                 = "Pooled"
  validate_environment = true
  load_balancer_type   = "BreadthFirst"

  registration_info {
    expiration_date = timeadd(timestamp(), "4h")
  }
}

output "new_hostpool_token" {
	value = azurerm_virtual_desktop_host_pool.test.registration_token
	sensitive = true
}

output "old_hostpool_token" {
	value = azurerm_virtual_desktop_host_pool.test.registration_info[0].token
	sensitive = true
}
```

New Output on second apply:

```
  # azurerm_virtual_desktop_host_pool.test will be updated in-place
  ~ resource "azurerm_virtual_desktop_host_pool" "test" {
        id                       = "/subscriptions/<redacted>/resourceGroups/acctestRG-vdesktop-1/providers/Microsoft.DesktopVirtualization/hostPools/acctestHP"
        name                     = "acctestHP"
      ~ registration_token       = (sensitive value)
        tags                     = {}
        # (8 unchanged attributes hidden)

      ~ registration_info {
          ~ expiration_date = "2021-11-12T21:25:37Z" -> (known after apply)
            # (2 unchanged attributes hidden)
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Changes to Outputs:
  # Warning: this attribute value will be marked as sensitive and will not
  # display in UI output after applying this change.
  ~ new_hostpool_token = (sensitive value)`
```

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
hey @amccullough84 

Thanks for this PR.

Taking a look through here whilst this is one approach to solving the problem, since the token has a timeframe based on the expiration date, in retrospect this probably makes less sense nested within the Virtual Desktop Host Pool resource and probably should be in it's own resource here, similar to:

```
resource "azurerm_virtual_desktop_host_pool_token" "test" {
  host_pool_id = azurerm_virtual_desktop_host_pool.test.id
  expiration_date = "..."
}

output "token" {
  value = azurerm_virtual_desktop_host_pool_token.test.token
}
```

Ultimately this would allow a similar fix to be applied to the top-level property `token` here, using the CustomizeDiff to mark it as being refreshed when the top level `expiration_date` field changes - and means we can remove the older inlined token in the next major release, so that users have a clear migration path (such that this won't impact any existing configurations) - WDYT?

Thanks!

**Comment by alexwilcox9** [comment]
Hey @tombuildsstuff 

With your proposed configuration where a token is its own resource what would happen if someone created two for the same hostpool? As a hostpool can only have one registration token at a time is there a way to detect this and handle it?

Also pretty sure this will solve an issue I raised so I'll link that here:
https://github.com/hashicorp/terraform-provider-azurerm/issues/13793

**Comment by amccullough84** [comment]
Excellent - that's a great approach @tombuildsstuff, much cleaner - I've re-worked that into a new azurerm_virtual_desktop_host_pool_registration_info resource which should meet the objective you outlined. Ready for review when you have a moment.

Acceptance Tests:

TF_ACC=1 go test -v ./internal/services/desktopvirtualization -run=TestAccVirtualDesktopHostPoolRegInfo -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccVirtualDesktopHostPoolRegInfo_basic
=== PAUSE TestAccVirtualDesktopHostPoolRegInfo_basic
=== RUN   TestAccVirtualDesktopHostPoolRegInfo_update
=== PAUSE TestAccVirtualDesktopHostPoolRegInfo_update
=== CONT  TestAccVirtualDesktopHostPoolRegInfo_basic
=== CONT  TestAccVirtualDesktopHostPoolRegInfo_update
--- PASS: TestAccVirtualDesktopHostPoolRegInfo_basic (89.51s)
--- PASS: TestAccVirtualDesktopHostPoolRegInfo_update (195.69s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/desktopvirtualization 195.843s

Have also updated the docs to include the new resource, and also add a note on the virtual_desktop_host_pool resource to advise use of the separate resource for managing registration info as using the inline option will eventually become inconsistent.

@alexwilcox9 - if duplicate resources - i.e. multiple resources created to set reg info on a hostpool are created, then one resource would be functional, one would potentially be inconsistent. I've added a note in the docs that only one should be specified for a given hostpool, but it feels like would be the expected behaviour - I can't see any valid reason why you would choose to specify two sets of registration info - other than misconfiguration.

**Comment by amccullough84** [comment]
Fixes #13793

**Review by jatinsshetty — APPROVED** [review]

**Comment by dschwinn** [comment]
I too am eagerly awaiting this release.  Thanks @amccullough84 for the work on this!

**Review by sfc-gh-gmacdougall — APPROVED** [review]
Please implement ASAP.

**Comment by alexwilcox9** [comment]
An update for everyone patiently waiting for this:
The main branch is now using a later API version so this PR needs to be updated to align with main, I thought this would be easy enough and gave it a go but encountered some issues. With the help of the folk in slack I've sorted most of them and there is now a PR on @amccullough84's branch [here](https://github.com/amccullough84/terraform-provider-azurerm/pull/1)

Still getting issues with tests failing on the `ExistsInAzure` check but hopefully that'll be sorted soon too

**Comment by amccullough84** [comment]
Thanks, will get into this later today

**Comment by dschwinn** [comment]
Thank you for working on this!!  It was a nice Happy New Year thing to see on Monday morning after a couple weeks off.

**Comment by amccullough84** [comment]
make acctests SERVICE='desktopvirtualization' TESTARGS='-run=TestAccVirtualDesktopHostPoolRegInfo' TESTTIMEOUT='30m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/desktopvirtualization -run=TestAccVirtualDesktopHostPoolRegInfo -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccVirtualDesktopHostPoolRegInfo_basic
=== PAUSE TestAccVirtualDesktopHostPoolRegInfo_basic
=== RUN   TestAccVirtualDesktopHostPoolRegInfo_update
=== PAUSE TestAccVirtualDesktopHostPoolRegInfo_update
=== CONT  TestAccVirtualDesktopHostPoolRegInfo_basic
=== CONT  TestAccVirtualDesktopHostPoolRegInfo_update
--- PASS: TestAccVirtualDesktopHostPoolRegInfo_basic (80.86s)
--- PASS: TestAccVirtualDesktopHostPoolRegInfo_update (175.48s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/desktopvirtualization 175.638s

**Comment by amccullough84** [comment]
Updated to include your changes @alexwilcox9, I've changed the acceptance tests so they're passing OK now - I can't see why the ExistsInAzure test fails - as you say, in actual use it works fine, but for some reason that test is failing. I've changed the test to verify that the "token" key exists in state after creation, which is ultimately the purpose of this anyway.

I've re-based from the current main branch as it's been sat around a while, so merging changes from upstream was making it hard to track the actual changes in the PR.

**Comment by danielmarker** [comment]
Up-voting this to be merged ASAP please - saves having to destroy 300 Vms to add an additional 2! Thanks all involved :)

**Comment by favoretti** [comment]
image
TC passes, code review from maintainers pending.

**Comment by dschwinn** [comment]
This is probably not the place to discuss such things but is there anywhere where we can see the roadmap for this project?  I've been watching this issue and waiting for the PR to merge literally since it the issue was created.  I too have deployments awaiting this release.  I've seen loads of other merges and release and I just wonder what the priorities are and why this has been so slow to be picked up.  Please don't take this as complaining. I am really just curious how the process works.  A lot of you do really great work, probably as volunteers so I really appreciate it.

And... I really will be happy when this merges.

**Comment by alexwilcox9** [comment]
I can only speculate as I'm not a maintainer of this repo nor do I work for Hashicorp but there is a pinned issue announcing the upcoming v3.0 release of the provider which I imagine is a lot of work. That and spinning back up after the holidays always takes a little while

They will get to it eventually however if in the meantime you want to try this functionality you can download the code and build a local provider, I wouldn't encourage using it in Prod as you might have to pick apart the state file should something in this PR change but it would be fine in a dev environment. Plus you might discover an issue the acceptance tests miss and can feedback here ahead of the review

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_registration_info_resource.go
Comment (tombuildsstuff) [inline internal/services/desktopvirtualization/virtual_desktop_host_pool_registration_info_resource.go:31]: presuming this has been copied & paste from the resource:

```suggestion
			Create: pluginsdk.DefaultTimeout(30 * time.Minute),
			Read:   pluginsdk.DefaultTimeout(5 * time.Minute),
			Update: pluginsdk.DefaultTimeout(30 * time.Minute),
			Delete: pluginsdk.DefaultTimeout(30 * time.Minute),
```

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go
Comment (tombuildsstuff) [inline internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go:22]: we also need to lock on this in the Create, Update and Delete methods of this resource?

Comment (tombuildsstuff) [reply internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go:22]: also I think `registration_info` needs to become computed in this resource, else there'll be a perpectual conflict?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @amccullough84 

Thanks for pushing those changes, apologies for the delayed re-review here.

I've taken a look through and left some comments inline, if we can fix those up then we should be able to take another look and get this in.

Thanks!

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go
Comment (amccullough84) [reply internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go:22]: > also I think `registration_info` needs to become computed in this resource, else there'll be a perpectual conflict?

The intent here was to avoid a breaking change - the docs for both the hostpool and the registration info make it clear  one or the other should be used - but not both, otherwise expect oddities!

WDYT? Is that enough or do we need to change to computed?

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_registration_info_resource.go
Comment (amccullough84) [reply internal/services/desktopvirtualization/virtual_desktop_host_pool_registration_info_resource.go]: @tombuildsstuff Sorry - not sure I follow this. The RetrieveRegistrationToken function returns a RegistrationInfo struct that doesn't contain a properties block?

I've re-worked the code to match your structure, but am confused about where we get the properties block from?

If there's no registration_info block set on the hostpool or the hostpool doesn't exist, we hit the error path on the function call. If we get back an empty block, we fail the ExpirationTime/Token check as they're the only two attributes with data.

Entirely possible I'm missing something obvious, so I'm open to correction!

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_registration_info_resource.go
Comment (amccullough84) [reply internal/services/desktopvirtualization/virtual_desktop_host_pool_registration_info_resource.go]: Resolved

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go
Comment (amccullough84) [reply internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go:23]: Added locks to `virtual_desktop_host_pool` resource

**Comment by amccullough84** [comment]
make acctests SERVICE='desktopvirtualization' TESTARGS='-run=TestAccVirtualDesktopHostPool' TESTTIMEOUT='30m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/desktopvirtualization -run=TestAccVirtualDesktopHostPool -timeout 30m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccVirtualDesktopHostPoolRegInfo_basic
=== PAUSE TestAccVirtualDesktopHostPoolRegInfo_basic
=== RUN   TestAccVirtualDesktopHostPoolRegInfo_update
=== PAUSE TestAccVirtualDesktopHostPoolRegInfo_update
=== RUN   TestAccVirtualDesktopHostPool_basic
=== PAUSE TestAccVirtualDesktopHostPool_basic
=== RUN   TestAccVirtualDesktopHostPool_complete
=== PAUSE TestAccVirtualDesktopHostPool_complete
=== RUN   TestAccVirtualDesktopHostPool_update
=== PAUSE TestAccVirtualDesktopHostPool_update
=== RUN   TestAccVirtualDesktopHostPool_requiresImport
=== PAUSE TestAccVirtualDesktopHostPool_requiresImport
=== CONT  TestAccVirtualDesktopHostPoolRegInfo_basic
=== CONT  TestAccVirtualDesktopHostPool_complete
=== CONT  TestAccVirtualDesktopHostPool_basic
=== CONT  TestAccVirtualDesktopHostPoolRegInfo_update
--- PASS: TestAccVirtualDesktopHostPoolRegInfo_basic (84.47s)
=== CONT  TestAccVirtualDesktopHostPool_requiresImport
--- PASS: TestAccVirtualDesktopHostPool_complete (88.09s)
=== CONT  TestAccVirtualDesktopHostPool_update
--- PASS: TestAccVirtualDesktopHostPool_basic (88.38s)
--- PASS: TestAccVirtualDesktopHostPoolRegInfo_update (182.97s)
--- PASS: TestAccVirtualDesktopHostPool_requiresImport (103.42s)
--- PASS: TestAccVirtualDesktopHostPool_update (169.48s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/desktopvirtualization 257.720s

**Comment by danielmarker** [comment]
Really appreciate @amccullough84 's commitment to this. Are we now in a position to merge a fix?

**Comment by danielmarker** [comment]
@tombuildsstuff  - are you happy this is now in a state to merge - so many users here wanting this to be merged  and a load of work by @amccullough84 who must be owed several beers for this!

**Comment by ChrisTav424** [comment]
We are having this problem in our organisation too. @amccullough84 thank you for fixing this. Do you know if there is a timescale?

**Review by mbfrahry — APPROVED** [review]
LGTM after following up on some of Tom's leftover comments!

File: internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go
Comment (tombuildsstuff) [reply internal/services/desktopvirtualization/virtual_desktop_host_pool_resource.go:23]: This will need to be computed once https://github.com/hashicorp/terraform-provider-azurerm/pull/14953 is merged - the latest API version changed the API behaviour from returning it in the original payload to using the secondary API

**Comment by gbeaud** [comment]
I am using registry.terraform.io/hashicorp/azurerm v2.95.0 but I still get errors when using both the old and new registration_token; do I do something wrong?

**Option 1:**

locals {
  registration_token = azurerm_virtual_desktop_host_pool.hostpool.registration_info[0].token
 }
 
│ Error: Invalid index
│
│   on host.tf line 6, in locals:
│    6:   registration_token = azurerm_virtual_desktop_host_pool.hostpool.registration_info[0].token
│     ├────────────────
│     │ azurerm_virtual_desktop_host_pool.hostpool.registration_info is empty list of object
│
│ The given key does not identify an element in this collection value: the collection has no elements.

**Option 2:**

locals {
 registration_token = azurerm_virtual_desktop_host_pool.hostpool.registration_token
}

│ Error: Unsupported attribute
│
│   on host.tf line 9, in locals:
│    9:   registration_token = azurerm_virtual_desktop_host_pool.hostpool.registration_token
│
│ This object has no argument, nested block, or exported attribute named "registration_token".
╵

Did anyone manage to use the new fix in their code? If yes could you please share your solution?

Thank you 😊

**Comment by amccullough84** [comment]
If you're using the resource based on the info in the docs:

```
resource "azurerm_resource_group" "example" {
  name     = "example-hostpool"
  location = "westeurope"
}

resource "azurerm_virtual_desktop_host_pool" "example" {
  name                 = "example-HP"
  location             = azurerm_resource_group.example.location
  resource_group_name  = azurerm_resource_group.example.name
  type                 = "Pooled"
  validate_environment = true
  load_balancer_type   = "BreadthFirst"

}

resource "azurerm_virtual_desktop_host_pool_registration_info" "example" {
  hostpool_id     = azurerm_virtual_desktop_host_pool.example.id
  expiration_date = "2022-01-01T23:40:52Z"
}
```

You would reference the token value using `azurerm_virtual_desktop_hostpool_registration_info.example.token`

Hope that helps!

**Pull request merged**
